File: apps/analytics/ml/plateau.py
"""Plateau detection via linear regression on weight progression.

For each exercise (slot_id), analyzes the last 8 weeks of successful sets
(min 8 data points). Fits linear regression on weekly-max weight.

Plateau: slope < 0.1 kg/week AND p-value > 0.1
Confidence: 1 - p_value, capped at 0.95
"""

from __future__ import annotations

from collections import defaultdict
from datetime import datetime, timedelta, timezone

import numpy as np
from scipy.stats import linregress

from queries import WorkoutRecord

_MIN_POINTS = 8
_WEEKS_BACK = 8
_PLATEAU_SLOPE_THRESHOLD = 0.1  # kg/week
_PLATEAU_PVALUE_THRESHOLD = 0.1
_MAX_CONFIDENCE = 0.95


def compute_per_exercise(records: list[WorkoutRecord]) -> dict[str, dict]:
    """Return slot_id -> plateau_detection payload for slots with enough data."""
    now = datetime.now(tz=timezone.utc)
    cutoff = now - timedelta(weeks=_WEEKS_BACK)

    by_slot: dict[str, list[WorkoutRecord]] = defaultdict(list)
    for r in records:
        if r.result != "success" or not r.recorded_at:
            continue
        try:
            dt = datetime.fromisoformat(r.recorded_at.replace("Z", "+00:00"))
        except ValueError:
            continue
        if dt >= cutoff:
            by_slot[r.slot_id].append(r)

    result: dict[str, dict] = {}
    for slot_id, slot_records in by_slot.items():
        payload = _analyze_slot(slot_records)
        if payload is not None:
            result[slot_id] = payload
    return result


def _analyze_slot(records: list[WorkoutRecord]) -> dict | None:
    if len(records) < _MIN_POINTS:
        return None

    by_week: dict[str, list[float]] = defaultdict(list)
    for r in records:
        dt = datetime.fromisoformat(r.recorded_at.replace("Z", "+00:00"))  # type: ignore[arg-type]
        cal = dt.isocalendar()
        week_key = f"{cal.year}-W{cal.week:02d}"
        by_week[week_key].append(r.weight)

    if len(by_week) < 2:
        return None

    weeks = sorted(by_week.keys())
    weights = [max(by_week[w]) for w in weeks]

    xs = np.arange(len(weeks), dtype=float)
    ys = np.array(weights, dtype=float)

    reg = linregress(xs, ys)
    slope = float(reg.slope)
    raw_pvalue = reg.pvalue
    # nan p_value means zero y-variance — perfectly flat, definitively a plateau
    pvalue_is_nan = raw_pvalue != raw_pvalue
    p_value = 0.0 if pvalue_is_nan else float(raw_pvalue)
    r_squared = 0.0 if pvalue_is_nan else float(reg.rvalue ** 2)

    is_plateau = slope < _PLATEAU_SLOPE_THRESHOLD and (pvalue_is_nan or p_value > _PLATEAU_PVALUE_THRESHOLD)
    # For nan case: perfectly flat → max confidence; otherwise confidence = 1 - p_value
    if is_plateau:
        confidence = _MAX_CONFIDENCE if pvalue_is_nan else min(1.0 - p_value, _MAX_CONFIDENCE)
    else:
        confidence = 0.0

    return {
        "isPlateauing": bool(is_plateau),
        "confidence": round(confidence, 3),
        "slope": round(slope, 3),
        "pValue": round(p_value, 4),
        "rSquared": round(r_squared, 3),
        "weeksAnalyzed": len(weeks),
        "currentWeight": weights[-1],
    }

File: apps/backend/api/src/analytics/pipelines/plateau.ts
/**
 * Plateau-detection insight (per exercise / slot_id).
 *
 * Ports apps/backend/analytics/ml/plateau.py. For each slot it takes the last
 * eight weeks of successful sets (min 8 points), fits a linear regression on
 * the weekly-max weight, and flags a plateau when the slope is below
 * 0.1 kg/week and the regression p-value is above 0.1.
 *
 * Degenerate (perfectly flat) series: scipy may report the p-value as NaN or as
 * 1.0 depending on version. The Agent A `linregress` port deterministically
 * returns p-value = 1.0 / r = 0 for a flat series (never NaN), so the degenerate
 * branch is detected here purely by `slope === 0 && popStd(weights) === 0`,
 * matching plateau.py's `slope == 0.0 and np.std(ys) == 0.0` fallback.
 */

import type { WorkoutRecord } from '../record';
import { linregress } from '../stats';
import { isoWeekKeyFromTimestamp } from '../iso-week';
import { parseInstant } from './time';
import { pyRound } from './round';

const MIN_POINTS = 8;
const WEEKS_BACK = 8;
const PLATEAU_SLOPE_THRESHOLD = 0.1; // kg/week
const PLATEAU_PVALUE_THRESHOLD = 0.1;
const MAX_CONFIDENCE = 0.95;
const MS_PER_WEEK = 7 * 86_400_000;

export interface PlateauDetectionPayload {
  readonly isPlateauing: boolean;
  readonly confidence: number;
  readonly slope: number;
  readonly pValue: number;
  readonly rSquared: number;
  readonly weeksAnalyzed: number;
  readonly currentWeight: number;
}

export function computePlateauPerExercise(
  records: readonly WorkoutRecord[]
): Map<string, PlateauDetectionPayload> {
  const cutoff = Date.now() - WEEKS_BACK * MS_PER_WEEK;

  const bySlot = new Map<string, WorkoutRecord[]>();
  for (const r of records) {
    if (r.result !== 'success' || r.recordedAt === null) continue;
    const instant = parseInstant(r.recordedAt);
    if (instant === null || instant.getTime() < cutoff) continue;
    const slot = bySlot.get(r.slotId) ?? [];
    slot.push(r);
    bySlot.set(r.slotId, slot);
  }

  const result = new Map<string, PlateauDetectionPayload>();
  for (const [slotId, slotRecords] of bySlot) {
    const payload = analyzeSlot(slotRecords);
    if (payload !== null) result.set(slotId, payload);
  }
  return result;
}

/**
 * Analyze a single slot's records (already filtered to the last eight weeks of
 * successful sets by `computePlateauPerExercise`). Exposed directly so parity
 * tests can drive it with fixed-timestamp fixtures, mirroring plateau.py's
 * `_analyze_slot`.
 */
export function analyzeSlot(records: readonly WorkoutRecord[]): PlateauDetectionPayload | null {
  if (records.length < MIN_POINTS) return null;

  const byWeek = new Map<string, number[]>();
  for (const r of records) {
    if (r.recordedAt === null) continue;
    const weekKey = isoWeekKeyFromTimestamp(r.recordedAt);
    if (weekKey === null) continue;
    const bucket = byWeek.get(weekKey) ?? [];
    bucket.push(r.weight);
    byWeek.set(weekKey, bucket);
  }

  if (byWeek.size < 2) return null;

  const weeks = [...byWeek.keys()].sort();
  const weights = weeks.map((w) => Math.max(...(byWeek.get(w) ?? [])));
  const xs = weeks.map((_, i) => i);

  const reg = linregress(xs, weights);
  const slope = reg.slope;
  const degenerate = slope === 0 && populationStd(weights) === 0;
  const pValue = degenerate ? 0 : reg.pValue;
  const rSquared = degenerate ? 0 : reg.rSquared;

  const isPlateau =
    slope < PLATEAU_SLOPE_THRESHOLD && (degenerate || pValue > PLATEAU_PVALUE_THRESHOLD);

  let confidence: number;
  if (!isPlateau) {
    confidence = 0;
  } else if (degenerate) {
    confidence = MAX_CONFIDENCE;
  } else {
    confidence = Math.min(1 - pValue, MAX_CONFIDENCE);
  }

  return {
    isPlateauing: isPlateau,
    confidence: pyRound(confidence, 3),
    slope: pyRound(slope, 3),
    pValue: pyRound(pValue, 4),
    rSquared: pyRound(rSquared, 3),
    weeksAnalyzed: weeks.length,
    currentWeight: weights[weights.length - 1],
  };
}

/** Population standard deviation (ddof=0), matching numpy `np.std`. */
function populationStd(values: readonly number[]): number {
  const n = values.length;
  let mean = 0;
  for (const v of values) mean += v;
  mean /= n;
  let sumSq = 0;
  for (const v of values) {
    const d = v - mean;
    sumSq += d * d;
  }
  return Math.sqrt(sumSq / n);
}
